example-pallet/src/lib.rs

```diff
@@ -0,0 +1,82 @@
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use chainbridge as bridge;
+use frame_support::traits::{Currency, ExistenceRequirement::AllowDeath, Get};
+use frame_support::{decl_event, decl_module, dispatch::DispatchResult};
+use frame_system::{self as system, ensure_signed, RawOrigin};
+use sp_runtime::traits::EnsureOrigin;
+use sp_std::prelude::*;
+
+mod mock;
+mod tests;
+
+type ResourceId = bridge::ResourceId;
+
+pub trait Trait: system::Trait + bridge::Trait {
+    type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+    /// Specifies the origin check provided by the bridge for calls that can only be called by the bridge pallet
+    type BridgeOrigin: EnsureOrigin<Self::Origin, Success = Self::AccountId>;
+
+    /// Ids can be defined by the runtime and passed in, perhaps from blake2b_128 hashes.
+    type HashId: Get<ResourceId>;
+    type NativeTokenId: Get<ResourceId>;
+}
+
+decl_event! {
+    pub enum Event<T> where
+        <T as frame_system::Trait>::Hash,
+    {
+        Remark(Hash),
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        const HashId: ResourceId = T::HashId::get();
+        const NativeTokenId: ResourceId = T::NativeTokenId::get();
+
+        fn deposit_event() = default;
+
+        //
+        // Initiation calls. These start a bridge transfer.
+        //
+
+        /// Transfers an arbitrary hash to a (whitelisted) destination chain.
+        pub fn transfer_hash(origin, hash: T::Hash, dest_id: bridge::ChainId) -> DispatchResult {
+            ensure_signed(origin)?;
+
+            let resource_id = T::HashId::get();
+            let metadata: Vec<u8> = hash.as_ref().to_vec();
+            <bridge::Module<T>>::transfer_generic(RawOrigin::Root.into(), dest_id, resource_id, metadata)
+        }
+
+        /// Transfers some amount of the native token to some recipient on a (whitelisted) destination chain.
+        pub fn transfer_native(origin, amount: u32, recipient: Vec<u8>, dest_id: bridge::ChainId) -> DispatchResult {
+            let source = ensure_signed(origin)?;
```

Comment:
Should `dest_id` not be checked here? If it is checked in `transfer_fungible` below only, a failure would leave the native tokens lost, since there is no state reverting in Substrate.